### PR TITLE
Optimized DispatchingEntityIdParser

### PR DESCRIPTION
**Pull request by mariushoch**
This made the function quite a bit faster:
> 14.48% 48797.778 817218 - Wikibase\DataModel\Entity\DispatchingEntityIdParser::parse

to

> 12.64% 42330.389 817218 - Wikibase\DataModel\Entity\DispatchingEntityIdParser::parse
when JSON-dumping 10,000 Wikidata entities.

Related to https://phabricator.wikimedia.org/T177486.

### CODE REVIEW

File: src/Entity/DispatchingEntityIdParser.php [hunk 1]
```diff
@@ -27,6 +27,10 @@ class DispatchingEntityIdParser implements EntityIdParser {
 	 * @param callable[] $idBuilders
 	 */
 	public function __construct( array $idBuilders ) {
+		if ( $this->idBuilders === [] ) {
+			throw new InvalidArgumentException( '$idBuilders must not be empty' );
```
Comment (thiemowmde) [inline src/Entity/DispatchingEntityIdParser.php]: Uh. I find this a little problematic. I believe this was done as it was before for a reason: The idea was to allow a setup with *no* custom entity types. As long as the parse method is not called, it is not a problem to have an empty ID parser.

Even if such a situation can never happen in production, this change makes this entire patch a breaking one. We could avoid this by keeping the previous handling of empty arrays.

The least minimum is to documented this as a `@throws`.

File: src/Entity/DispatchingEntityIdParser.php [hunk 2]
```diff
@@ -37,20 +41,19 @@ public function __construct( array $idBuilders ) {
 	 * @return EntityId
 	 */
 	public function parse( $idSerialization ) {
-		$this->assertIdIsString( $idSerialization );
+		try {
+			list( , , $localId ) = EntityId::splitSerialization( $idSerialization );
+		} catch ( InvalidArgumentException $ex ) {
+			// Only do this assert in case something is wrong, EntityId::splitSerialization already
+			// checks for string.
+			$this->assertIdIsString( $idSerialization );
```
Comment (thiemowmde) [inline src/Entity/DispatchingEntityIdParser.php]: Why keep this additional check *inside* the catch? This looks awkward. What does this do in addition to what splitSerialization already does? Let's remove it, please.

File: src/Entity/DispatchingEntityIdParser.php (hunk 1 above)
Comment (mariushoch) [reply src/Entity/DispatchingEntityIdParser.php]: Moved back, this probably hardly has any performance impact. It mostly looked weird to me.

File: src/Entity/DispatchingEntityIdParser.php (hunk 2 above)
Comment (mariushoch) [reply src/Entity/DispatchingEntityIdParser.php]: changing it will change the Exception message slightly, but I guess that's ok.

**Comment by mariushoch** [comment]
Note: This is no longer breaking

**Pull request merged**